actors/builtin/miner/miner_test.go

```diff
@@ -390,6 +390,112 @@ func TestProvingPeriodCron(t *testing.T) {
+		}
+
+		func() {
+			defer func() {
+				r := recover()
+				require.NotNil(t, r)
+				assert.Contains(t, fmt.Sprintf("%s",r), "cannot reduce sector expiration")
+			}()
+
+			actor.extendSector(rt, sector, 0, params)
+		}()
+	})
+
+	t.Run("rejects extention to invalid epoch", func(t *testing.T) {
+		rt, _, _, sector := setupExtension(t, builder, actor, periodOffset)
+
+		// attempt to extend to an epoch that is not a multiple of the proving period + the commit epoch
+		extension := 42*miner.WPoStProvingPeriod + 1
+		newExpiration := sector.Info.Expiration - abi.ChainEpoch(extension)
+		params := &miner.ExtendSectorExpirationParams{
+			SectorNumber:  sector.Info.SectorNumber,
+			NewExpiration: newExpiration,
+		}
+
+		func() {
+			defer func() {
+				r := recover()
+				require.NotNil(t, r)
+				assert.Contains(t, fmt.Sprintf("%s",r), "invalid expiration")
+			}()
+
+			actor.extendSector(rt, sector, uint64(extension), params)
+		}()
+	})
+
+	t.Run("updates expiration with valid params", func(t *testing.T) {
+		rt, st, expiration, sector := setupExtension(t, builder, actor, periodOffset)
+
+		extension := uint64(42 * miner.WPoStProvingPeriod)
+		newExpiration := sector.Info.Expiration + abi.ChainEpoch(extension)
+		params := &miner.ExtendSectorExpirationParams{
+			SectorNumber:  sector.Info.SectorNumber,
+			NewExpiration: newExpiration,
+		}
+
+		actor.extendSector(rt, sector, extension, params)
+
+		// assert sector expiration is set to the new value
+		st = getState(rt)
+		sector, found, err := st.GetSector(rt.AdtStore(), sector.Info.SectorNumber)
+		require.NoError(t, err)
+		require.True(t, found)
+		assert.Equal(t, newExpiration, sector.Info.Expiration)
+
+		// assert that an expiration exists at the target epoch
+		expirations, err := st.GetSectorExpirations(rt.AdtStore(), newExpiration)
+		require.NoError(t, err)
+		exists, err := expirations.IsSet(uint64(sector.Info.SectorNumber))
+		require.NoError(t, err)
+		assert.True(t, exists)
+
+		// assert that the expiration has been removed from the old epoch
+		expirations, err = st.GetSectorExpirations(rt.AdtStore(), expiration)
+		require.NoError(t, err)
+		exists, err = expirations.IsSet(uint64(sector.Info.SectorNumber))
+		require.NoError(t, err)
+		assert.False(t, exists)
+	})
+}
+
+func setupExtension(t *testing.T, b *mock.RuntimeBuilder, a *actorHarness, periodOffset abi.ChainEpoch) (*mock.Runtime, *miner.State, abi.ChainEpoch, *miner.SectorOnChainInfo) {
+	rt := b.Build(t)
+	a.constructAndVerify(rt, periodOffset)
+	precommitEpoch := abi.ChainEpoch(1)
+	rt.SetEpoch(precommitEpoch)
+	st := getState(rt)
+	deadline := st.DeadlineInfo(rt.Epoch())
+	expiration := deadline.PeriodEnd() + 10*miner.WPoStProvingPeriod
+	sectorInfo := a.commitAndProveSectors(rt, 1, expiration, big.Zero())
+
+	sector, found, err := getState(rt).GetSector(rt.AdtStore(), sectorInfo[0].SectorNumber)
+	require.NoError(t, err)
+	require.True(t, found)
+	return rt, st, expiration, sector
+}
```

Comment:
I don't love this approach, because it makes these setup functions easy to abuse as a blunt way to set up other, unrelated tests, which leads to unwanted coupling between the tests and general maintenance pain. This is based on my experience working with existing uses of the pattern in go-filecoin.

I think there probably is room to factor out some generic set-up code more than is already done, but I would want to do so in a way that is explicitly not coupled to any particular tests, probably some kind of builder pattern.

In this case, I'd prefer to share the setup code by just testing multiple conditions in the one test case. E.g. see TestCommitSector.